src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -108,23 +131,107 @@ Score(const size_t queryIndex, TreeType& referenceNode)
 
     // Calculate kernel value based on reference node centroid.
     if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
-    {
       kernelValue = EvaluateKernel(queryIndex, referenceNode.Point(0));
-    }
     else
-    {
-      kde::KDEStat& referenceStat = referenceNode.Stat();
       kernelValue = EvaluateKernel(queryPoint, referenceStat.Centroid());
-    }
 
     densities(queryIndex) += referenceNode.NumDescendants() * kernelValue;
 
     // Don't explore this tree branch.
     score = DBL_MAX;
+
+    // Store not used alpha for Monte Carlo.
+    if (kernelIsGaussian && monteCarlo)
+      accumMCAlpha(queryIndex) += depthAlpha;
+  }
+  else if (monteCarlo &&
+           referenceNode.NumDescendants() >= mcAccessCoef * initialSampleSize &&
+           newCalculations &&
+           kernelIsGaussian)
+  {
+    // Monte Carlo probabilistic estimation.
+    // Calculate z using accumulated alpha if possible.
+    const double alpha = depthAlpha + accumMCAlpha(queryIndex);
+    const boost::math::normal normalDist;
+    const double z =
+        std::abs(boost::math::quantile(normalDist, alpha / 2));
+
+    // Auxiliary variables.
+    const size_t numDesc = referenceNode.NumDescendants();
+    arma::vec sample;
+    size_t m = initialSampleSize;
+    double meanSample;
+    bool useMonteCarloPredictions = true;
+
+    // Resample as long as confidence is not high enough.
+    while (m > 0)
+    {
+      const size_t oldSize = sample.size();
+      const size_t newSize = oldSize + m;
+
+      // Don't use probabilistic estimation if this is going to take a similar
+      // amount of computations to the exact calculation.
+      if (newSize >= mcBreakCoef * numDesc)
+      {
+        useMonteCarloPredictions = false;
+        break;
+      }
+
+      // Increase the sample size.
+      sample.resize(newSize);
+      for (size_t i = 0; i < m; ++i)
+      {
+        // Sample and evaluate random points from the reference node.
+        const size_t randomPoint = math::RandInt(0, numDesc);
+        sample(oldSize + i) =
+            EvaluateKernel(queryIndex, referenceNode.Descendant(randomPoint));
+      }
+      meanSample = arma::mean(sample);
+      const double stddev = arma::stddev(sample);
+      const double mThreshBase =
+          z * stddev * (1 + relError) / (relError * meanSample);
+      const size_t mThresh = std::ceil(mThreshBase * mThreshBase);
+
+      if (sample.size() < mThresh)
+        m = mThresh - sample.size();
+      else
+        m = 0;
+    }
+
+    if (useMonteCarloPredictions)
+    {
+      // Confidence is high enough so we can use Monte Carlo estimation.
+      densities(queryIndex) += numDesc * meanSample;
+      score = DBL_MAX;
+
+      // Accumulated alpha has been used.
+      accumMCAlpha(queryIndex) = 0;
+    }
+    else
+    {
+      // Recurse.
+      score = minDistance;
+
+      if (referenceNode.IsLeaf() || tree::TreeTraits<TreeType>::HasSelfChildren)
```

Comment:
It is exactly as you you wrote, `tree::TreeTraits<TreeType>::HasSelfChildren` means that the tree is a cover tree. But the point is cover trees just have single or dual traversers but not breadth first right?
```
template<typename RuleType>
using BreadthFirstDualTreeTraverser = DualTreeTraverser<RuleType>;
```
This is from `cover_tree.hpp`

Reply:
The default cover tree traverser is somewhat strange: it does a depth-first traversal for the query tree, but it does a breadth-first traversal for the reference tree (this is kind of weird, but remember it can be both of these things at once since we're recursing in both the query tree and the reference tree in a dual-tree recursion).  So for the needs of the `BreadthFirstDualTreeTraverser`, the default `CoverTree::DualTreeTraverser` is sufficient.